File: src/lerobot/gui/api/live_video.py
"""The Run tab's live camera video: one endpoint a viewer offers itself to.

A viewer opens its connection as soon as the tab is shown, before the run has
frames — an offer with no video streams is answered with the cycle channel
alone — and offers again on the same connection when the cameras appear. The
first picture then costs a keyframe and one way across the link, rather than
a connection's whole setup.

The pipeline runs only while someone is watching a camera: the first offer
that asks for streams starts it, the last viewer's departure stops it, so a
run nobody is watching pays nothing. Every viewer shares the one pipeline and
its one encoder per camera; each gets its own connection, its own
subscription and its own packets.
"""

from __future__ import annotations

import asyncio
import logging
import secrets
from concurrent.futures import ThreadPoolExecutor
from dataclasses import asdict, dataclass, field

from aiortc import RTCConfiguration, RTCPeerConnection, RTCSessionDescription
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from lerobot.gui.link_class import CLASS_LINK
from lerobot.gui.live_video.pipeline import PROFILE_WIDTH, CycleMessage, LivePipeline
from lerobot.gui.live_video.transport import CYCLE_CHANNEL, LiveVideoSession, video_codecs_chosen

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/run/live-video", tags=["live-video"])

#: Starting and stopping the pipeline join threads, so they wait off the
#: event loop — in their own pool, never the shared one a stalled job
#: elsewhere could fill.
_lifecycle = ThreadPoolExecutor(max_workers=1, thread_name_prefix="live-video-lifecycle")


@dataclass
class _Viewer:
    """One page's connection, from before the run until it goes away."""

    connection: RTCPeerConnection
    #: None until the viewer asks for cameras; the pipeline's subscription
    #: and the tracks on it.
    session: LiveVideoSession | None = None
    cameras: list[str] = field(default_factory=list)
    #: The cycle channel the page opened, which it does in its first offer —
    #: before any session exists to hear the event.
    channel: object | None = None


_pipeline: LivePipeline | None = None
_viewers: dict[str, _Viewer] = {}
_lock: asyncio.Lock | None = None
_lock_loop: asyncio.AbstractEventLoop | None = None


def _viewers_lock() -> asyncio.Lock:
    """The lock every path that reads or replaces the pipeline holds.

    Built on first use rather than at import: a lock belongs to one event
    loop, and one built at import belongs to whichever loop happens to touch
    it first — which is not necessarily the loop the server ends up running
    on, and never the same one twice across a suite.
    """
    global _lock, _lock_loop
    loop = asyncio.get_running_loop()
    if _lock is None or _lock_loop is not loop:
        _lock, _lock_loop = asyncio.Lock(), loop
    return _lock


class Offer(BaseModel):
    sdp: str
    type: str
    #: The connection this offer belongs to, from a previous answer. Absent
    #: for the first offer, which opens one.
    session: str | None = None
    #: Whether this offer's video streams are meant to carry cameras. The
    #: offer that opens a connection before the run has any sets it false:
    #: its one video stream is a placeholder, there because every later
    #: stream shares the transport chosen for the first one ([O20]).
    attach: bool = True


def pipeline_running() -> bool:
    return _pipeline is not None


def _watching() -> int:
    return sum(1 for v in _viewers.values() if v.session is not None)


def _reader_or_none():
    from lerobot.robots.obs_stream import ObservationStreamReader

    try:
        return ObservationStreamReader()
    except Exception:
        return None


def _profile_for(cameras: int) -> dict:
    """What the page shows beside the stream's state, and what it is against."""
    return {
        "profile": {
            "width": PROFILE_WIDTH,
            "bitrate_kbit_s": CLASS_LINK.per_camera_kbit_s(cameras) if cameras else None,
        },
        "link": {
            "name": CLASS_LINK.name,
            "down_kbit_s": CLASS_LINK.down_kbit_s,
            "rtt_ms": CLASS_LINK.rtt_ms,
            "stream_budget_kbit_s": CLASS_LINK.stream_budget_kbit_s,
        },
    }


def _resolve_pose(state: dict[str, float]) -> dict:
    """The visualizer tile's joint angles for one cycle's state.

    The same shape the tile's own endpoint returns, so the tile applies it
    unchanged — it is fed from the stream instead of polling, which is what
    keeps it as fresh as the pictures. A robot with no vendored description
    is not an error: the tile is simply not shown.
    """
    from lerobot.gui.urdf_viz import compute_joint_angles, resolve_robot

    spec = resolve_robot(state.keys())
    if spec is None:
        return {"available": False}
    angles = compute_joint_angles(spec, state)
    return {
        "available": True,
        "arms": [
            {"prefix": a.obs_prefix, "frames": [{"joints": angles.get(a.obs_prefix, {})}]} for a in spec.arms
        ],
    }


def _message_for_the_page(message: CycleMessage) -> dict:
    body = asdict(message)
    body["pose"] = _resolve_pose(message.state)
    return body


@router.get("/status")
async def status() -> dict:
    """What the controls bar shows: whether there is anything to watch, at
    what profile, and how many people are watching."""
    reader = _pipeline or _reader_or_none()
    if reader is None:
        return {"available": False, "cameras": [], "viewers": 0, "encoders": 0}
    cameras = list(reader.cameras if _pipeline is not None else reader.image_keys)
    if _pipeline is None:
        reader.close()
    return {
        "available": True,
        "cameras": cameras,
        "viewers": _watching(),
        "encoders": _pipeline.encoder_count if _pipeline is not None else 0,
        **_profile_for(len(cameras)),
        "cameras_detail": _pipeline.snapshot() if _pipeline is not None else {},
    }


def _build_and_start() -> LivePipeline:
    pipeline = LivePipeline()
    # Starting pays the device's one-time costs before a frame is asked
    # for, so the first viewer's first picture is not the one that waits.
    pipeline.start()
    return pipeline


async def _ensure_pipeline() -> LivePipeline:
    global _pipeline
    if _pipeline is not None and not _pipeline.reads_the_current_tap():
        # The run it was built for has ended. Its reader holds segments
        # nobody writes to any more, so serving from it would be a stream
        # that never advances again — indistinguishable, on the page, from
        # a run standing still.
        logger.info("live video: the run this pipeline read has ended, building one for the new tap")
        pipeline, _pipeline = _pipeline, None
        await asyncio.get_running_loop().run_in_executor(_lifecycle, pipeline.stop)
    if _pipeline is None:
        try:
            # Building asks the device which encoders it has, which is the
            # first CUDA call in this process and costs seconds of it: the
            # loop this runs on is the one serving every other request.
            pipeline = await asyncio.get_running_loop().run_in_executor(_lifecycle, _build_and_start)
        except FileNotFoundError as e:
            # No tap: there is genuinely no run, which is the ordinary state
            # between runs and not a fault. This is the answer the page treats
            # as "nothing to watch yet" rather than as a failure.
            logger.info("live video: no run to watch (%s)", e)
            raise HTTPException(503, "No run is streaming, so there is nothing to watch yet") from e
        except Exception as e:
            # A run IS streaming and the pipeline would not start -- no encoder,
            # a device that refused, a camera the reader cannot map. Answering
            # "nothing to watch" sends the operator to look at the robot while
            # the only account of what happened sits in a log line nobody is
            # reading. The design requires this state to carry a reason.
            logger.exception("live video: the pipeline would not start")
            raise HTTPException(503, f"The live stream could not start: {type(e).__name__}: {e}") from e
        _pipeline = pipeline
        logger.info("live video: pipeline started for %s", pipeline.cameras)
    return _pipeline


async def _stop_pipeline_if_nobody_is_watching() -> None:
    """Stop the pipeline when the last viewer of it has gone.

    The pipeline is started by the first offer that attaches cameras, so a
    viewer that never becomes one — a refused offer, a connection that died
    during its own negotiation — would otherwise leave it encoding for
    nobody, with no viewer left to leave and stop it.
    """
    global _pipeline
    if _watching() == 0 and _pipeline is not None:
        pipeline, _pipeline = _pipeline, None
        await asyncio.get_running_loop().run_in_executor(_lifecycle, pipeline.stop)


def _check_offer(wanted: list[str], pipeline: LivePipeline) -> None:
    """What an offer asking for cameras must look like.

    An answer cannot add a video stream the offer did not describe, and
    cannot change the codec the offer put first, so an offer wrong in either
    way would be answered with a stream that carries nothing the viewer can
    decode. Say so instead.
    """
    if len(wanted) != len(pipeline.cameras):
        raise HTTPException(
            400,
            f"This run has {len(pipeline.cameras)} cameras and the offer describes "
            f"{len(wanted)} video streams. Ask /status for the cameras, then offer one "
            "receive-only video transceiver for each.",
        )
    wrong = sorted({c for c in wanted if c.upper() != "H264"})
    if wrong:
        raise HTTPException(
            400,
            f"This stream is H.264; the offer asks for {', '.join(wrong)} first. "
            "Put H.264 first on each video transceiver before offering.",
        )


@router.post("/offer")
async def offer(body: Offer) -> dict:
    """Answer a viewer's offer, on a new connection or the one it names."""
    async with _viewers_lock():
        wanted = video_codecs_chosen(body.sdp)
        # `attach` is the page saying which of its two moves this is, so an
        # offer that claims to carry cameras and describes none is wrong in
        # the same way as one that describes too few.
        attaching = body.attach

        if body.session is None:
            # No ICE servers: the page reaches this server directly — the
            # same host, the LAN, or the tailnet — so host candidates are the
            # whole answer, and asking a public one costs seconds of setup
            # for a candidate nothing here uses.
            token = secrets.token_urlsafe(12)
            connection = RTCPeerConnection(RTCConfiguration(iceServers=[]))
            viewer = _Viewer(connection=connection)
            _viewers[token] = viewer

            @connection.on("datachannel")
            def on_channel(channel) -> None:
                if channel.label == CYCLE_CHANNEL:
                    viewer.channel = channel

            @connection.on("connectionstatechange")
            async def on_state_change() -> None:
                if connection.connectionState in ("failed", "closed", "disconnected"):
                    await _drop(token)
        else:
            token = body.session
            viewer = _viewers.get(token)
            if viewer is None:
                raise HTTPException(404, "That session is not open here; offer without one to open it")
            connection = viewer.connection

        if attaching:
            pipeline = await _ensure_pipeline()
            try:
                _check_offer(wanted, pipeline)
            except HTTPException:
                await _stop_pipeline_if_nobody_is_watching()
                raise
        elif wanted and len(wanted) > 1:
            raise HTTPException(
                400,
                "An offer that is not attaching cameras carries one placeholder video "
                f"stream, not {len(wanted)}.",
            )

        await connection.setRemoteDescription(RTCSessionDescription(sdp=body.sdp, type=body.type))
        if attaching and viewer.session is None:
            # The session reads the remote description to learn which side
            # opened the cycle channel, so it is built after the offer is
            # applied — and only now, because a connection with no cameras
            # has nothing to subscribe to.
            viewer.session = LiveVideoSession(
                pipeline, connection, decorate=_message_for_the_page, channel=viewer.channel
            )
            viewer.cameras = viewer.session.cameras
        await connection.setLocalDescription(await connection.createAnswer())
        # Which stream is which camera: the page has no way to tell from the
        # description, and the order is this session's, not the status
        # endpoint's — a run that restarts with a camera unplugged would
        # otherwise relabel every tile.
        return {
            "sdp": connection.localDescription.sdp,
            "type": connection.localDescription.type,
            "cameras": viewer.cameras,
            "session": token,
            # The profile the tiles are at, so the page can say what it is
            # showing without a second request for it.
            **_profile_for(len(viewer.cameras)),
        }


async def _drop(token: str) -> None:
    """Close one viewer's connection and stop the pipeline if it was the last.

    Takes the same lock the offer path holds: both read and replace the
    pipeline, and a viewer leaving while another is negotiating would
    otherwise take the pipeline out from under it.
    """
    async with _viewers_lock():
        await _drop_held(token)


async def _drop_held(token: str) -> None:
    """Precondition: the caller holds the viewers lock."""
    viewer = _viewers.pop(token, None)
    if viewer is None:
        return
    if viewer.session is not None:
        await viewer.session.close()
    await viewer.connection.close()
    logger.info("live video: a viewer left, %d watching", _watching())
    await _stop_pipeline_if_nobody_is_watching()


async def shutdown() -> None:
    """Close every viewer and stop the pipeline. For server shutdown and tests."""
    global _pipeline
    async with _viewers_lock():
        for token in list(_viewers):
            await _drop_held(token)
        # A pipeline with nobody watching is normally stopped by the last
        # drop; stopping it here too means shutdown holds whatever happened.
        if _pipeline is not None:
            pipeline, _pipeline = _pipeline, None
            await asyncio.get_running_loop().run_in_executor(_lifecycle, pipeline.stop)

File: src/lerobot/gui/static/live_video.js
// The Run tab's live camera video: the page's side of the stream.
// Design: src/lerobot/gui/docs/live_camera_video.md
//
// One connection carries a video track per camera and one data channel with
// the cycle's state, action and robot pose. The page makes the offer and the
// server answers it, so everything the stream needs has to be in the offer:
// one receive-only stream per camera, H.264 first on each (our frames are
// H.264 and the answering side cannot change that), and the cycle channel.
// The answer names the cameras in the order of its streams, which is the
// only thing that says which tile is which.
//
// Age at the eye: a painted frame's timestamp is its capture time in the
// transport's clock plus a constant the library picks at random and never
// sends. The channel announces every cycle's capture time, so the constant
// that turns the most readings into times that were announced is the one;
// readings no constant explains are reported as an age we do not know.
(() => {
    const CLOCK = 90000;           // RTP's clock for video, ticks per second
    const WRAP = 2 ** 32;          // the timestamp is 32 bits
    const RECENT = 120;            // readings the age is taken over: a few seconds
    const ANNOUNCED = 600;         // capture times kept for the join: ~20 s at 30 fps
    const RELEARN_AFTER = 30;      // unexplained frames before the constant is searched again

    function rtpFromCapture(seconds) {
        return Math.round(seconds * CLOCK) % WRAP;
    }

    function createAgeTracker() {
        const announced = [];                 // capture times, newest last
        const announcedSet = new Set();       // their timestamps, for the join
        const perCamera = new Map();          // camera -> {origin, readings:[{wire, shownAt}]}

        function noteCapture(captureTs) {
            announced.push(captureTs);
            announcedSet.add(rtpFromCapture(captureTs));
            while (announced.length > ANNOUNCED) {
                announcedSet.delete(rtpFromCapture(announced.shift()));
            }
        }

        function learnOrigin(readings) {
            // Try each of the newest few readings against every announced
            // capture time and keep the constant that explains the most. A
            // wrong one explains almost none, because real capture times are
            // irregular at the clock's resolution — but a perfectly regular
            // cadence is explained equally well by a constant a whole period
            // out, and then there is no answer to give rather than one that
            // would read as an age.
            let best = null;
            let bestVotes = 0;
            let tied = false;
            const probes = readings.slice(-8);
            for (const probe of probes) {
                for (const stamp of announcedSet) {
                    const candidate = ((probe.wire - stamp) % WRAP + WRAP) % WRAP;
                    if (candidate === best) continue;
                    let votes = 0;
                    for (const r of readings) {
                        if (announcedSet.has(((r.wire - candidate) % WRAP + WRAP) % WRAP)) votes++;
                    }
                    if (votes > bestVotes) { best = candidate; bestVotes = votes; tied = false; }
                    else if (votes === bestVotes && candidate !== best) { tied = true; }
                }
            }
            if (tied) return null;
            return bestVotes >= Math.max(3, Math.floor(readings.length / 4)) ? best : null;
        }

        function explains(origin, wire) {
            return announcedSet.has(((wire - origin) % WRAP + WRAP) % WRAP);
        }

        function notePainted(camera, wireRtp, shownAtSeconds) {
            let entry = perCamera.get(camera);
            if (!entry) { entry = { origin: null, readings: [], misses: 0 }; perCamera.set(camera, entry); }
            entry.readings.push({ wire: wireRtp, shownAt: shownAtSeconds });
            if (entry.readings.length > RECENT) entry.readings.shift();
            // The constant holds for the life of a track, so the search runs
            // once. A re-offered track is a new constant, and it shows as
            // frames that the current one no longer explains: a run of those
            // is what asks for the search again, rather than a clock. The
            // check is one lookup; the search is every announced time
            // against every reading, and running it per frame would cost the
            // page more than the pictures do.
            if (entry.origin !== null) {
                entry.misses = explains(entry.origin, wireRtp) ? 0 : entry.misses + 1;
                if (entry.misses < RELEARN_AFTER) return;
            }
            const learned = learnOrigin(entry.readings);
            if (learned !== null) { entry.origin = learned; entry.misses = 0; }
        }

        function captureOf(wire, origin, near) {
            const span = WRAP / CLOCK;
            const base = near - (near % span);
            const ticks = ((wire - origin) % WRAP + WRAP) % WRAP;
            let best = null;
            for (const k of [-1, 0, 1]) {
                const t = base + ticks / CLOCK + k * span;
                if (best === null || Math.abs(t - near) < Math.abs(best - near)) best = t;
            }
            return best;
        }

        function medianAgeMs(camera) {
            const entry = perCamera.get(camera);
            if (!entry || entry.origin === null || !announced.length) return null;
            const near = announced[announced.length - 1];
            const ages = [];
            for (const r of entry.readings) {
                ages.push((r.shownAt - captureOf(r.wire, entry.origin, near)) * 1000);
            }
            if (!ages.length) return null;
            ages.sort((a, b) => a - b);
            return ages[Math.floor(ages.length / 2)];
        }

        return { noteCapture, notePainted, medianAgeMs };
    }

    function defaultVideoCodecs() {
        if (typeof RTCRtpSender === 'undefined' || !RTCRtpSender.getCapabilities) return [];
        const caps = RTCRtpSender.getCapabilities('video');
        return (caps && caps.codecs) || [];
    }

    function createClient(opts = {}) {
        const doFetch = opts.fetch || ((...a) => fetch(...a));
        const PeerConnection = opts.RTCPeerConnection
            || (typeof RTCPeerConnection !== 'undefined' ? RTCPeerConnection : null);
        const videoCodecs = opts.videoCodecs || defaultVideoCodecs;
        const onState = opts.onState || (() => {});
        const onTrack = opts.onTrack || (() => {});
        const onCycle = opts.onCycle || (() => {});
        const tracker = createAgeTracker();

        const client = {
            state: { name: 'idle', reason: '' },
            cameras: [],
            profile: null,
            // The newest track per camera, so a tile rebuilt while the
            // stream is up is given what is already arriving.
            tracks: {},
            open,
            attach,
            close,
            notePainted,
            ageMs: (camera) => tracker.medianAgeMs(camera),
        };
        let pc = null;
        let session = null;
        let transceivers = [];
        let closed = false;
        // One negotiation at a time. Opening early and attaching the cameras
        // are started from different places — the tab appearing, the run's
        // frames appearing — and a connection can only carry one offer at a
        // time: overlapping them leaves the second answering a description
        // the first has already replaced.
        let chain = Promise.resolve();

        function serialize(work) {
            const next = chain.then(work, work);
            chain = next.catch(() => {});
            return next;
        }

        function setState(name, reason = '') {
            // Opening and attaching both pass through connecting, and a
            // listener that redraws on every call would redraw for a state
            // that did not change.
            if (client.state.name === name && client.state.reason === reason) return;
            client.state = { name, reason };
            onState(client.state);
        }

        // Set while the bar's failure is the server's rather than the
        // connection's, so a recovery clears the one and not the other.
        let failedOnCameras = false;

        function noteFailing(failing) {
            const cameras = Object.keys(failing || {});
            if (!cameras.length) {
                if (failedOnCameras) { failedOnCameras = false; setState('connecting'); }
                return;
            }
            failedOnCameras = true;
            setState('failed', cameras.map((c) => `${c}: ${failing[c]}`).join('; '));
        }

        function notePainted(camera, wireRtp, shownAtSeconds) {
            tracker.notePainted(camera, wireRtp, shownAtSeconds);
            if (['connecting', 'idle'].includes(client.state.name)) setState('streaming');
        }

        function h264Codecs() {
            return videoCodecs().filter((c) => (c.mimeType || '').toLowerCase() === 'video/h264');
        }

        function addVideoStream(codecs) {
            const t = pc.addTransceiver('video', { direction: 'recvonly' });
            if (t.setCodecPreferences) t.setCodecPreferences(codecs);
            transceivers.push(t);
            return t;
        }

        async function negotiate(attaching) {
            const offer = await pc.createOffer();
            await pc.setLocalDescription(offer);
            const response = await doFetch('/api/run/live-video/offer', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    sdp: pc.localDescription.sdp,
                    type: pc.localDescription.type,
                    session,
                    attach: attaching,
                }),
            });
            if (!response.ok) {
                let reason = `the server refused the offer (${response.status})`;
                try { reason = (await response.json()).detail || reason; } catch (_) { /* keep it */ }
                throw new Error(reason);
            }
            const answer = await response.json();
            session = answer.session || session;
            if (answer.cameras && answer.cameras.length) client.cameras = answer.cameras;
            if (answer.profile) client.profile = answer.profile;
            await pc.setRemoteDescription(answer);
            return answer;
        }

        // Open the connection before the run has frames, so the first
        // picture costs a keyframe and one way across the link rather than a
        // connection's whole setup. The offer carries one placeholder video
        // stream: every later stream shares the transport chosen for the
        // first one, so a connection opened with the channel alone could
        // only ever carry one camera.
        async function _open() {
            if (pc) return true;
            // WebRTC first: the codec list comes from an RTCRtpSender, so a
            // browser without WebRTC reports no H.264 either and would be
            // told the wrong thing about itself.
            if (!PeerConnection) {
                setState('failed', 'this browser has no WebRTC');
                return false;
            }
            const h264 = h264Codecs();
            if (!h264.length) {
                setState('failed', 'this browser cannot decode H.264, which is what this stream is');
                return false;
            }
            setState('connecting');
            // `closed` says this client was closed for good, not that a
            // connection once was: a new one clears it, or the first
            // failed open would silence every failure after it.
            closed = false;
            const connection = new PeerConnection({ iceServers: [] });
            pc = connection;
            const channel = pc.createDataChannel('cycles');
            channel.onmessage = (e) => {
                let message;
                try { message = JSON.parse(e.data); } catch (_) { return; }
                if (typeof message.capture_ts === 'number') tracker.noteCapture(message.capture_ts);
                noteFailing(message.failing);
                onCycle(message);
            };
            addVideoStream(h264);
            pc.ontrack = (e) => {
                const index = transceivers.indexOf(e.transceiver);
                const camera = client.cameras[index];
                if (!camera) return;
                client.tracks[camera] = e.track;
                onTrack(camera, e.track);
            };
            connection.onconnectionstatechange = () => {
                if (closed || pc !== connection) return;
                // A browser's "disconnected" is a link that has gone quiet,
                // and it comes back: on a link that drops for seconds at a
                // time, calling it a failure would leave the bar reading
                // failed for the rest of a run that recovered.
                if (connection.connectionState === 'disconnected') {
                    if (client.state.name === 'streaming') setState('connecting');
                    return;
                }
                if (['failed', 'closed'].includes(connection.connectionState)) {
                    setState('failed', `the connection ${connection.connectionState}`);
                }
            };
            try {
                await negotiate(false);
                // Open, and carrying nothing: the operator opened the tab
                // before launching. Saying "connecting" here would describe
                // a connection that is already up and read as a fault.
                if (!client.cameras.length) setState('idle');
                return true;
            } catch (e) {
                setState('failed', e && e.message ? e.message : String(e));
                close();
                return false;
            }
        }

        // The run's cameras exist: bind them to the connection already up.
        async function _attach(cameras) {
            if (!pc && !(await _open())) return;
            const carrying = client.cameras;
            const same = carrying.length === cameras.length && carrying.every((c, i) => c === cameras[i]);
            if (same && Object.keys(client.tracks).length) {
                return;  // already carrying these
            }
            // A connection's video streams only ever grow, and their order
            // is what names them. A run that restarts with a camera gone —
            // or with the same count under other names — cannot be carried
            // by this connection: the offer would describe more streams than
            // the run has, or label every tile with the wrong camera.
            if (carrying.length && !same) {
                close();
                if (!(await _open())) return;
            }
            const h264 = h264Codecs();
            while (transceivers.length < cameras.length) addVideoStream(h264);
            setState('connecting');
            try {
                await negotiate(true);
            } catch (e) {
                setState('failed', e && e.message ? e.message : String(e));
            }
        }

        function open() {
            return serialize(_open);
        }

        function attach(cameras) {
            return serialize(() => _attach(cameras));
        }

        function close() {
            closed = true;
            if (pc) { try { pc.close(); } catch (_) { /* already gone */ } }
            pc = null;
            session = null;
            transceivers = [];
            client.tracks = {};
        }

        return client;
    }

    window.LiveVideo = { createClient, createAgeTracker, rtpFromCapture, CLOCK, WRAP };
})();
